crates/oxc_linter/src/rules/eslint/no_useless_call.rs

```diff
@@ -0,0 +1,207 @@
+use oxc_ast::{
+    ast::{CallExpression, ChainElement, Expression, MemberExpression},
+    match_member_expression, AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::cmp::ContentEq;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_useless_call_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Avoid unnecessary use of .call() and .apply()")
+        .with_help("Replace with a normal function invocation")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUselessCall;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow unnecessary calls to .call() and .apply()
+    ///
+    /// ### Why is this bad?
+    /// Function.prototype.call() and Function.prototype.apply() are slower than the normal function invocation.
+    ///
+    /// ### Examples
+    ///
+    /// ```js
+    /// // These are same as `foo(1, 2, 3);`
+    /// foo.call(undefined, 1, 2, 3);
+    /// foo.apply(undefined, [1, 2, 3]);
+    /// foo.call(null, 1, 2, 3);
+    /// foo.apply(null, [1, 2, 3]);
+    ///
+    /// // These are same as `obj.foo(1, 2, 3);`
+    /// obj.foo.call(obj, 1, 2, 3);
+    /// obj.foo.apply(obj, [1, 2, 3]);
+    /// ```
+    ///
+    /// ```js
+    /// // The `this` binding is different.
+    /// foo.call(obj, 1, 2, 3);
+    /// foo.apply(obj, [1, 2, 3]);
+    /// obj.foo.call(null, 1, 2, 3);
+    /// obj.foo.apply(null, [1, 2, 3]);
+    /// obj.foo.call(otherObj, 1, 2, 3);
+    /// obj.foo.apply(otherObj, [1, 2, 3]);
+    ///
+    /// // The argument list is variadic.
+    /// // Those are warned by the `prefer-spread` rule.
+    /// foo.apply(undefined, args);
+    /// foo.apply(null, args);
+    /// obj.foo.apply(obj, args);
+    /// ```
+    NoUselessCall,
+    eslint,
+    perf
+);
+
+impl Rule for NoUselessCall {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else { return };
+
+        let Some(callee) =
+            as_member_expression_without_chain_expression(call_expr.callee.without_parentheses())
+        else {
+            return;
+        };
+
+        if !is_call_or_non_variadic_apply(call_expr, callee) {
+            return;
+        }
+
+        let applied = callee.object().without_parentheses();
+        let Some(first_arg) = call_expr.arguments.first() else { return };
+        let Some(this_arg) = first_arg.as_expression() else { return };
+
+        if validate_this_argument(this_arg, applied) {
+            ctx.diagnostic(no_useless_call_diagnostic(call_expr.span));
+        }
+    }
+}
+
+fn validate_this_argument<'a>(this_arg: &Expression<'a>, applied: &Expression<'a>) -> bool {
```

Comment:
I find it hard to get the lifetimes right. When the lifetime is not needed, I would like to see it removed:

```suggestion
fn validate_this_argument(this_arg: &Expression, applied: &Expression) -> bool {
```

The same for the other outside functions :)

Reply:
I’ve removed the unnecessary lifetime in the commit below!
https://github.com/oxc-project/oxc/pull/8789/commits/15b7cec252548170156898ac6df40e48ac8a578a

However, for the `as_member_expression_without_chain_expression()`, the lifetime specification was kept because deleting it would cause an error.

Reply:
Because its only one param in and one param out, I do not see a problem to keep this. Thanks you 👍